0000-encryption-api/README.md

```diff
@@ -0,0 +1,135 @@
+
+# Summary
+[summary]: #summary
+
+Encryption is considered a primary use for cryptography which Ursa does not
+currently provide. There is demand for providing easy to use encryption/decryption
+interfaces in Ursa. Ursa already has similar interfaces for signing/verification.
+The implemented algorithms provide programmers choices that aim
+to avoid bad parameter interactions. Keys are passed around as data blobs
+allowing consumers to protect them as they see fit.
+
+# Motivation
+[motivation]: #motivation
+
+Hyperledger projects will need to be able to encrypt/decrypt data using
+industry standards. To minimize mistakes and allow other projects good
+encryption schemes, Ursa will provide known good implementations with
+good APIs that are simple enough to avoid confusion.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Ursa should distinguish between two methods for encryption: one for asymmetric algorithms
+like RSA, and one for symmetric like AES. However the interfaces will be very similar.
+
+[Libsodium](https://libsodium.gitbook.io/doc/) for example uses these two types as public-key encryption and secret key encryption.
+
+Public key encryption is generally applied for key exchange or secret key wrapping. It requires two different keys–*public* and *private*, linked together with a one way mathematical
+relationship. The *private* key is used to decrypt a message and cannot be recovered from the *public* key and MUST remain private. The *public* key is used to encrypt a message.
+
+Secret key encryption uses the same key to encrypt and decrypt. It is only shared among all involved parties as a *shared secret key*.
+
+Users will create keys using one of these two types and pass them to the appropriate method.
+Rust typing will prevent a user from passing a secret key to a public key encryption method and visa versa.
+
+Public-Key encryption methods MUST use the naming convention **ALGORITHM or CURVE{SIZE}-SCHEME**. Below are some examples
+that could be implemented:
+
+- Rsa2048Oaep
+    - Encryption with an RSA 2048 bit public key using [Optimal Asymmetric Encryption](https://tools.ietf.org/html/rfc8017#section-7.1).
+    - Decryption with an RSA 2048 bit private key.
+    - This method does not use a MAC or authentication tag.
+- X25519Chacha20Poly1305
+    - Encryption with shared secret derived from X25519 public key and ephemeral X25519 private key using CHACHA-POLY1305.
+    - Decryption with shared secret derived from X25519 private key and ephemeral X25519 public key using CHACHA-POLY1305.
+    - Uses authentication tag.
+
+Secret-key encryption methods MUST use the naming convention **ALGORITHM{SIZE}{-MAC}**. Below are some examples
+that could be implemented:
+
+- Aes256Gcm
+- Chacha20Poly1305
+- Twofish256CbcHmacSha2_256
+- Blowfish448CbcHmacSha2_256
+
+Secret-key encryption methods require the use of *nonces* or *initialization vectors (iv)*, the two terms are often used interchangeably.
```

Comment:
initialization vectors are different from nonce but do we want to talk about IVs at all

Reply:
I understand they are different but in many newer crypto libraries they use both. Yes we should mention them to prevent ambiguity.